### DISCUSSION TITLE
Oriented Bounding Box (OBB) Datasets Overview - Ultralytics YOLOv8 Docs

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#5378] [Docs]
# Oriented Bounding Box (OBB) Datasets Overview - Ultralytics YOLOv8 Docs

Dive deep into various oriented bounding box (OBB) dataset formats compatible with Ultralytics YOLO models. Grasp the nuances of using and converting datasets to this format.

https://docs.ultralytics.com/datasets/obb/

**Comment by AnirAnz456346** [comment]
Hi

I tried to use object counter given at https://docs.ultralytics.com/guides/object-counting/ 

It seems that currently OBB models don't support tracking as I'm getting this error.

NotImplementedError: ERROR ❌ OBB task does not support track mode!

If compatibility could be added for same that would have a lot of applications.

**Reply by RizwanMunawar** [reply]
@AnirAnz456346 Appreciate you sharing the details of the issue you faced. Indeed, you're correct that object tracking is not currently supported with the OBB module. We have plans to release this feature in the coming days. Stay tuned.

Thanks,
Ultralytics Team!

**Reply by scastagnetta** [reply]
Hi @RizwanMunawar, do you have updates on this? Do you have an idea of when this will be available? I am also willing to contribute if needed.

**Comment by karthikdatta98** [comment]
Can we use SAHI on YOLO OBB?

**Reply by RizwanMunawar** [reply]
@karthikdatta98 OBB support with SAHI is not available yet, we may provide this support in the future.

Thanks
Ultralytics Team!

**Reply by pao444** [reply]
Hello, I was just wondering if there was any update on using YOLO OBB models with SAHI, and if not if you might have any guidance on how it could be implemented

**Reply by glenn-jocher** [reply]
Hey there! 😊 As of now, there isn't direct support for using YOLO OBB models with SAHI, but it's a great suggestion! Implementing this could involve converting YOLO OBB predictions to a format that SAHI understands. While we don't have a ready-made solution, you might explore converting the YOLO OBB output (the corner points or the `xywhr` format) into a polygon or a rotated bounding box format that SAHI can process.

For a quick start, here's a rough idea on how you might begin converting YOLO OBB outputs for SAHI:

```python
# Example: Convert YOLO OBB output to a format SAHI understands
# Let's say you have YOLO OBB output as [class_index, x1, y1, x2, y2, x3, y3, x4, y4]

yolo_obb_output = [0, 0.780811, 0.743961, 0.782371, 0.74686, 0.777691, 0.752174, 0.776131, 0.749758]

# Convert this to your desired format, e.g., a polygon that SAHI might accept
sahi_polygon = [(yolo_obb_output[i], yolo_obb_output[i+1]) for i in range(1, len(yolo_obb_output), 2)]

# Now, sahi_polygon contains polygon points that you might use with SAHI
```

This is quite simplified and would need adjustments based on your exact requirements and the SAHI API. Keep an eye on our updates; we're always working on enhancing compatibility and features based on community feedback!

If you dive into this, we'd love to hear about your progress or any insights you might have. Good luck! 🚀

**Reply by pao444** [reply]
Hey @glenn-jocher! Thanks for your quick and informative reply! I'll give it a shot and keep you in the loop with my progress. I appreciate you taking the time out of your day to help! Cheers!

**Reply by glenn-jocher** [reply]
Hey there! 🌟 I'm thrilled to hear you're giving it a go and really appreciate you keeping us updated on your journey. Your kind words mean a lot! Remember, if you hit any bumps along the way or have more questions, feel free to reach out. We're here to help. Happy coding and best of luck with your project! Cheers!

**Comment by karthikyerram** [comment]
can we use the yolov8 txt annotation format for this? If not, is there a way to convert?

**Reply by pderrenger** [reply]
@karthikyerram yes, you can use the YOLOv8 txt annotation format for oriented bounding boxes (OBB). The YOLO OBB format specifies bounding boxes by their four corner points with coordinates normalized between 0 and 1, following the format: `class_index, x1, y1, x2, y2, x3, y3, x4, y4`.

If your annotations are not already in this format and you need to convert them, you can use the provided conversion script in the Ultralytics YOLOv8 documentation. The script `convert_dota_to_yolo_obb` is designed to transition labels from the DOTA dataset format to the YOLO OBB format, which is compatible with the Ultralytics YOLO models.

Here's a quick overview of how you can prepare and convert your dataset:

- Ensure your dataset annotations are in the correct YOLO OBB format.
- If you have annotations in a different format, such as the standard DOTA format, use the provided conversion script to align them with the YOLO OBB format.
- Once converted, validate that the annotations match the expected structure.
- Update your dataset's YAML configuration file with the correct paths, classes, and class names.

By following these steps, you can effectively train your YOLOv8 model with OBB annotations. 🚀

**Comment by pax7** [comment]
this is great news!

1. is the format for the mark-ups like before, where the first 4 numbers are the corners of the rectangle, or is now 5 numbers, the rectangle corners and an angle?

2. can we have OBB-pose soon? Ideally we would also have segment-pose but OBB-pose would be amazing too.

**Reply by pderrenger** [reply]
@pax7 1. The OBB format in YOLOv8 indeed uses 8 numbers representing the four corners of the rectangle, normalized between 0 and 1. The format is `class_index, x1, y1, x2, y2, x3, y3, x4, y4`. Internally, YOLO processes these as `xywhr` (center x, center y, width, height, rotation), but the annotation format remains with the corners specified.

2. The idea of combining OBB with pose estimation (OBB-pose) is intriguing! 🤔 While we currently don't have this feature, it's certainly something that could be considered for future updates. The integration of pose estimation with OBB would indeed be a significant advancement. For now, we continue to improve and support the existing models and features. Stay tuned for any updates on new capabilities like OBB-pose! 🚀

If you have more questions or need further assistance, feel free to check out our documentation at https://docs.ultralytics.com. 😊

**Comment by Rashid-7864** [comment]
Use of Computer Vision in order to Identify the damages on Car (Cost Estimation)
Dear ultralytics Team please support me 

I am working for an insurance company, currently we are following the manual process for surveying the damaged car in order to undertsand the Damage and Estimate the Insurance cost accordingly

I am looking for a Smarter way to identify the Damage ,along with part of the card
I am looking forward to create a YOLOv8 intence segmentataion which will help business in identifying insurance cost in a fater and effcient way

Below is the work flow

Customer uploads the Car image
Model Identify the car model and company name 
Model Identify te parts of the car ( Door, Bumper, Headlight,Rear-view mirror,Hood,Roof Etc)
Model Identify the Level of Damage ( High , Medium and Low)
Model creates the report and shows overall estimation of the insurance
Please help me with anyrelevant examples or codes which i refer to build my model
why ultralytics team can't responded?!

**Reply by pderrenger** [reply]
@Rashid-7864 hello! It's great to hear that you're considering using computer vision and YOLOv8 for automating damage assessment and cost estimation for car insurance. Here's how you can approach this task with YOLOv8:

1. **Dataset Preparation**: Collect a dataset of car images with various damages. Annotate the images with bounding boxes around the damaged areas and label them with the car part and damage level.

2. **Model Selection**: Choose a suitable YOLOv8 model architecture. For instance, YOLOv8n (nano) for faster inference or YOLOv8x (extra) for higher accuracy.

3. **Custom Training**: Train a YOLOv8 model on your annotated dataset. You'll need to create a custom dataset YAML file that specifies the paths to your images and labels, as well as the class names.

4. **Inference and Analysis**: Once the model is trained, you can use it to predict on new images. The model will identify car parts and damage levels, which you can then use to estimate insurance costs.

5. **Report Generation**: Develop a system to compile the model's predictions into a comprehensive report, including the estimated cost.

Here's a simplified example of how you might train a YOLOv8 model with a custom dataset:

```python
from ultralytics import YOLO

# Create a new YOLO model from scratch or load a pretrained one
model = YOLO('yolov8n.yaml')  # From scratch
# model = YOLO('yolov8n.pt')  # Pretrained

# Train the model on your custom dataset
results = model.train(data='your_dataset.yaml', epochs=100)
```

For validation:

```python
# Validate the model's performance on the validation set
results = model.val(data='your_dataset.yaml')
```

For inference:

```python
# Perform inference on new images
results = model.predict('path/to/car_image.jpg')

# Process results
for result in results:
    print(result.boxes)  # Bounding boxes
    print(result.boxes.cls)  # Predicted classes
    # Additional processing to create the report
```

Please note that building a model for damage assessment is a complex task that may require fine-tuning and iterative development. You might also need to integrate additional logic to translate the model's outputs into cost estimates.

As for your concern about not receiving a response, please understand that the team might be handling a high volume of inquiries or working on updates. However, I'm here to help guide you through the process. If you have further questions or need more detailed guidance, feel free to ask! 🚗💡

**Comment by tom123478** [comment]
I was wondering if you provided some specific protocol instructions for yolov8-obb? For example, the angle definition method used, and the use of independent detection heads for angles. I would be grateful if you could tell me, as that would save me a lot of time.

**Reply by pderrenger** [reply]
@tom123478 absolutely, I'd be happy to clarify how YOLOv8 handles Oriented Bounding Boxes (OBB):

- **Angle Definition**: In YOLOv8, the angle for OBB is defined within the model's architecture. The model predicts the rotation of the bounding box, which is represented as an angle in radians. This angle is used to calculate the four corners of the OBB, providing a more precise localization of objects, especially in aerial or satellite imagery where objects can be oriented in any direction.

- **Detection Heads**: YOLOv8 utilizes a dedicated detection head for angle prediction within the network. This means that alongside the traditional bounding box coordinates and class predictions, there's a specific part of the model responsible for predicting the rotation angle of each detected object.

- **Training with OBB**: When training YOLOv8 with OBB, the model expects the dataset to be in a specific OBB format. This format includes the four corner points of the bounding box, normalized between 0 and 1. The model then internally processes these as `xywhr` (center x, center y, width, height, rotation) for loss calculations and outputs.

To ensure the best results with OBB:

- Make sure your dataset is properly formatted to the YOLO OBB format before training.
- If you're using a dataset like DOTA, you may need to convert the annotation format to be compatible with YOLOv8.
- Check the documentation for any additional configuration settings that might be specific to OBB training.

I hope this helps streamline your process! If you need more detailed guidance, the Ultralytics documentation is a comprehensive resource for all things related to YOLOv8 and OBB. 📘🚀

**Comment by dansojka** [comment]
Hello, 

Any plans (or is it possible) to add the "Track" mode to OBB?

**Reply by glenn-jocher** [reply]
@dansojka yes, this is now live in the latest version. Happy tracking with OBB :)

**Reply by pderrenger** [reply]
@dansojka hello!

Currently, the "Track" mode in Ultralytics YOLOv8 is designed for axis-aligned bounding boxes (AABB). However, integrating "Track" mode with Oriented Bounding Boxes (OBB) is an interesting idea that could potentially enhance tracking accuracy, especially in scenarios where objects are rotated or skewed.

While there are no immediate plans announced for this feature, it's certainly possible to consider for future updates. The YOLO community and the Ultralytics team are always working on improvements and new features, so stay tuned for updates! If you have a specific use case or ideas for this feature, your feedback is always welcome. 🚀

In the meantime, if you're comfortable with coding, you could experiment with adapting the current tracking algorithms to work with OBBs, or keep an eye on the repository for any community contributions in this area. 🛠️

Remember, the success of open-source projects like YOLOv8 often relies on community involvement and contributions. So, if you or anyone else is interested in developing this feature, it could be a valuable addition to the project! 🌟

**Comment by DmitriyBerbencev** [comment]
How can I get the coordinates of the oriented bounding box?

**Reply by AnirAnz456346** [reply]
You can do it by using prediction results.
For example

results = model.predict(image_source)
r = results[0]

for box in r.obb:
    cords = box.xywhr[0].tolist() # [x, y, width, height, theta (in degrees E [0,90])]
    x = cords[0]
    y = cords[1]
   w = cords[2]
   h = cords[3]
   theta = cords[4]

here x,y are coordinates of box centroid, w,h are bounding box width and height, and theta is orientation
If you're working on a video source, you'll need an extra for loop eg for r in results

**Reply by DmitriyBerbencev** [reply]
thnks, you saved me

**Reply by glenn-jocher** [reply]
@AnirAnz456346 yes this looks correct!

**Reply by shilpiprd06** [reply]
Hello, 
So, I understand that the yolov8-obb input format is 8 numbers which essentially represent the coordinates of the bounding box. But since the model internally processes these as xywhr, and the output provided is in this format as well, can't we provide input in this format as well ? I'd really appreciate if you could tell me soon, as this would save me a lot of time.

**Reply by glenn-jocher** [reply]
@shilpiprd06 hello,

Currently, the YOLOv8 models are designed to accept input in the YOLO OBB format (the 8 coordinates format) for training. While the model does internally convert these to the `xywhr` format for processing, the input format needs to be in the specified 8 coordinates to ensure compatibility with the preprocessing steps.

If converting your data to this format is a concern, you might consider using the provided script `convert_dota_to_yolo_obb` to automate the conversion of your dataset from DOTA format to YOLO OBB format. This should help streamline your workflow.

Hope this helps! 😊

**Reply by shilpiprd06** [reply]
Thanks @glenn-jocher  for texting back and clearing my doubts !

**Reply by glenn-jocher** [reply]
You're welcome! 😊 For more detailed information on Oriented Bounding Box (OBB) datasets and their formats, you can refer to the [Oriented Bounding Box (OBB) Datasets Overview](https://docs.ultralytics.com/datasets/obb/). This guide covers the YOLO OBB format, supported datasets, and how to convert label formats. If you have any more questions, feel free to ask!

**Comment by hariouat** [comment]
Hello,
To ensure my annotations are correctly formatted for YOLO-OBB, I have initial annotations in the format class_id, X1, Y1, X2, y2, X3, Y3, X4, Y4. I then converted these annotations to the DOTA format: x1, y1, x2, y2, x3, y3, x4, y4, category, difficult. However, when using the convert_dota_to_yolo_obb function, it doesn't seem to recognize my new classes. Could you please guide me on how to add these new classes and successfully execute the function?

**Reply by glenn-jocher** [reply]
Hi @hariouat,

The `convert_dota_to_yolo_obb` function has a hardcoded class mapping dictionary that only contains the standard DOTA dataset classes (plane, ship, storage-tank, etc.). When you try to use it with your custom classes, they're not recognized because they don't exist in this predefined dictionary.

Here are a couple of solutions you could try:

1. Create a custom conversion script by copying and modifying the `convert_dota_to_yolo_obb` function from the source code. You can update the class_mapping dictionary to include your own classes:

```python
from pathlib import Path
import cv2
from tqdm import tqdm as TQDM

def convert_custom_dota_to_yolo_obb(dota_root_path):
    dota_root_path = Path(dota_root_path)
    
    # Replace with your own class mapping
    class_mapping = {
        "your_class_1": 0,
        "your_class_2": 1,
        # Add all your classes here
    }
    
    def convert_label(image_name, image_width, image_height, orig_label_dir, save_dir):
        orig_label_path = orig_label_dir / f"{image_name}.txt"
        save_path = save_dir / f"{image_name}.txt"
        
        with orig_label_path.open("r") as f, save_path.open("w") as g:
            lines = f.readlines()
            for line in lines:
                parts = line.strip().split()
                if len(parts) < 9:
                    continue
                class_name = parts[8]
                if class_name in class_mapping:
                    class_idx = class_mapping[class_name]
                    coords = [float(p) for p in parts[:8]]
                    normalized_coords = [
                        coords[i] / image_width if i % 2 == 0 else coords[i] / image_height for i in range(8)
                    ]
                    formatted_coords = [f"{coord:.6g}" for coord in normalized_coords]
                    g.write(f"{class_idx} {' '.join(formatted_coords)}\n")
    
    # Rest of the function implementation
    for phase in ["train", "val"]:
        image_dir = dota_root_path / "images" / phase
        orig_label_dir = dota_root_path / "labels" / f"{phase}_original"
        save_dir = dota_root_path / "labels" / phase
        
        save_dir.mkdir(parents=True, exist_ok=True)
        
        image_paths = list(image_dir.iterdir())
        for image_path in TQDM(image_paths, desc=f"Processing {phase} images"):
            if image_path.suffix != ".png":
                continue
            image_name_without_ext = image_path.stem
            img = cv2.imread(str(image_path))
            h, w = img.shape[:2]
            convert_label(image_name_without_ext, w, h, orig_label_dir, save_dir)
```

2. Alternatively, if you already have annotations in the format `class_id, X1, Y1, X2, Y2, X3, Y3, X4, Y4`, you might not need the converter at all. You just need to ensure the class IDs match your dataset's YAML file classes and that the coordinates are normalized between 0 and 1.

For more information on OBB datasets, check out our [documentation page](https://docs.ultralytics.com/datasets/obb/).

**Reply by hariouat** [reply]
> Hi @hariouat,
> 
> The `convert_dota_to_yolo_obb` function has a hardcoded class mapping dictionary that only contains the standard DOTA dataset classes (plane, ship, storage-tank, etc.). When you try to use it with your custom classes, they're not recognized because they don't exist in this predefined dictionary.
> 
> Here are a couple of solutions you could try:
> 
>     1. Create a custom conversion script by copying and modifying the `convert_dota_to_yolo_obb` function from the source code. You can update the class_mapping dictionary to include your own classes:
> 
> 
> ```python
> from pathlib import Path
> import cv2
> from tqdm import tqdm as TQDM
> 
> def convert_custom_dota_to_yolo_obb(dota_root_path):
>     dota_root_path = Path(dota_root_path)
>     
>     # Replace with your own class mapping
>     class_mapping = {
>         "your_class_1": 0,
>         "your_class_2": 1,
>         # Add all your classes here
>     }
>     
>     def convert_label(image_name, image_width, image_height, orig_label_dir, save_dir):
>         orig_label_path = orig_label_dir / f"{image_name}.txt"
>         save_path = save_dir / f"{image_name}.txt"
>         
>         with orig_label_path.open("r") as f, save_path.open("w") as g:
>             lines = f.readlines()
>             for line in lines:
>                 parts = line.strip().split()
>                 if len(parts) < 9:
>                     continue
>                 class_name = parts[8]
>                 if class_name in class_mapping:
>                     class_idx = class_mapping[class_name]
>                     coords = [float(p) for p in parts[:8]]
>                     normalized_coords = [
>                         coords[i] / image_width if i % 2 == 0 else coords[i] / image_height for i in range(8)
>                     ]
>                     formatted_coords = [f"{coord:.6g}" for coord in normalized_coords]
>                     g.write(f"{class_idx} {' '.join(formatted_coords)}\n")
>     
>     # Rest of the function implementation
>     for phase in ["train", "val"]:
>         image_dir = dota_root_path / "images" / phase
>         orig_label_dir = dota_root_path / "labels" / f"{phase}_original"
>         save_dir = dota_root_path / "labels" / phase
>         
>         save_dir.mkdir(parents=True, exist_ok=True)
>         
>         image_paths = list(image_dir.iterdir())
>         for image_path in TQDM(image_paths, desc=f"Processing {phase} images"):
>             if image_path.suffix != ".png":
>                 continue
>             image_name_without_ext = image_path.stem
>             img = cv2.imread(str(image_path))
>             h, w = img.shape[:2]
>             convert_label(image_name_without_ext, w, h, orig_label_dir, save_dir)
> ```
> 
>     2. Alternatively, if you already have annotations in the format `class_id, X1, Y1, X2, Y2, X3, Y3, X4, Y4`, you might not need the converter at all. You just need to ensure the class IDs match your dataset's YAML file classes and that the coordinates are normalized between 0 and 1.
> 
> 
> For more information on OBB datasets, check out our [documentation page](https://docs.ultralytics.com/datasets/obb/).

Thank you for your replay. please it's important the points orders?

**Reply by glenn-jocher** [reply]
@hariouat Yes, the order of points is important when working with OBB annotations in YOLO. For consistent results, you should arrange the four corner points in a clockwise or counter-clockwise order. While YOLO is somewhat flexible in processing these points (as it converts them to the `xywhr` format internally), maintaining a consistent ordering helps ensure accurate training and inference.

When creating your annotations, I recommend following this approach:
1. Arrange points either clockwise or counter-clockwise
2. Ideally start from a consistent position (e.g., top-left corner)
3. Ensure all your annotations follow the same ordering pattern

This consistency is especially important when working with custom datasets or when converting from other formats. The model will learn better when the point order follows a predictable pattern across all training examples.
